### PR TITLE
Iss 252 as validator i want to cross check my study data against the study definexml for missing datasets so that i can stay compliant with sd1063

### PR DESCRIPTION
**Pull request by htu**
1. Download the following test files:
[combined_datasets.json.txt](https://github.com/cdisc-org/cdisc-rules-engine/files/11899150/combined_datasets.json.txt)
[define.xml.txt](https://github.com/cdisc-org/cdisc-rules-engine/files/11899152/define.xml.txt)
[rule_sd1063a.json.txt](https://github.com/cdisc-org/cdisc-rules-engine/files/11899154/rule_sd1063a.json.txt)
[rule_sd1063b.json.txt](https://github.com/cdisc-org/cdisc-rules-engine/files/11899155/rule_sd1063b.json.txt)

2. Put the files in a folder on your local computer such as ./dist/issue252

3. Remove the file extension .txt from file names

4. Run the test cases: 
```python -m pytest -s tests/unit/test_dataset_builders/test_contents_define_dataset_builder```
You should see 10 cases Passed

5. Run the rule test A
```
"Check": {
        "all": [
            {
                "name": "dataset_name",
                "operator": "not_equal_to",
                "value": "define_dataset_name"
            }
        ]
    },
    "Rule_Type": "Dataset Metadata Check against Define XML"
}
```

```python core.py test -s sdtmig -v 3-4 -dp ./dist/issue252/combined_datasets.json -r ./dist/issue252/rule_sd1063a.json -dv 2.1 ```
Check the report: 1) the TS domain should not show in the report; 2) all the other domains should be reported  

6. Run the rule test B
```
   "Check": {
        "any": [
            {
                "name": "dataset_name",
                "operator": "empty"
            },
            {
                "name": "define_dataset_name",
                "operator": "empty"
            }
        ]
    },
    "Rule_Type": "Dataset Metadata Check against Define XML"
}
``` 
```python core.py test -s sdtmig -v 3-4 -dp ./dist/issue252/combined_datasets.json -r ./dist/issue252/rule_sd1063b.json -dv 2.1 ```
Check the report: 1) the TS domain should not show in the report; 2) all the other domains should be reported  

7. Run a regresion test 
```python core.py validate -s sdtmig -v 3-4 -d ./dist/issue252 ```

### CODE REVIEW

File: cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        define_col_order = [
+            "define_dataset_name",
+            "define_dataset_label",
+            "define_dataset_location",
+            "define_dataset_class",
+            "define_dataset_structure",
+            "define_dataset_is_non_standard",
+            "define_dataset_variables",
+        ]
+        # define_df = pd.DataFrame(kwargs.get("define_metadata", {}))
```
Comment (gerrycampion) [inline cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py:44]: remove comment

File: cdisc_rules_engine/rules_engine.py [hunk 2]
```diff
@@ -178,30 +185,41 @@ def get_dataset_builder(
...
+        kwargs = {}
         builder = self.get_dataset_builder(rule, dataset_path, datasets, domain)
         dataset = builder.get_dataset()
-        kwargs = {}
+        define_metadata = self.define_metadata
+
         # Update rule for certain rule types
         # SPECIAL CASES FOR RULE TYPES ###############################
         # TODO: Handle these special cases better.
         if (
```
Comment (gerrycampion) [inline cdisc_rules_engine/rules_engine.py:210]: I think this logic should be done in the builder so that we can get rid of this section altogether and so that we don't need to pass around kwargs and define_metadata everywhere.
Refer to `define_item_group_dataset_builder`.
You could make the `domain` parameter optional in `get_define_xml_item_group_metadata`

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py [hunk 3]
```diff
@@ -338,7 +346,7 @@ def validate_schema(self) -> bool:
         logger.info(f"Validated Define-XML schema. is_valid={is_valid}")
         return is_valid
 
-    @cache
+    @cached
```
Comment (gerrycampion) [inline cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:349]: Why did you change this caching?

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py [hunk 4]
```diff
@@ -197,6 +198,13 @@ def _get_domain_metadata(self, metadata, domain_name):
                 f"Domain {domain_name} is not found in Define XML"
             )
 
+    def _get_all_metadata_items(self, metadata):
```
Comment (gerrycampion) [inline cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:201]: is this being used anywhere?

**Review by gerrycampion — CHANGES_REQUESTED** [review]

File: tests/unit/test_dataset_builders/test_contents_define_dataset_builder.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        domain=None,
+        define_metadata=define_metadata,
+    ).build(**kwargs)
+    exp_columns = result.columns.tolist()
+    req_columns = [
+        "dataset_size",
+        "dataset_location",
+        "dataset_name",
+        "dataset_label",
+        #    "variables",
```
Comment (gerrycampion) [inline tests/unit/test_dataset_builders/test_contents_define_dataset_builder.py:367]: Remove comment

File: cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py (hunk 1 above)
Comment (htu) [reply cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py:44]: Done.

File: cdisc_rules_engine/rules_engine.py (hunk 2 above)
Comment (htu) [reply cdisc_rules_engine/rules_engine.py:210]: I created a method get_dataset_define_metadata in the base_dataset_builder, which does not use **kwargs. We check the rule type here. How would you like this part to be improved?

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py (hunk 3 above)
Comment (htu) [reply cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:349]: Thought that it was cached before somewhere and forgot to look for the cached name. Changed it back.

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py (hunk 4 above)
Comment (htu) [reply cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:201]: No. Remove it?

File: tests/unit/test_dataset_builders/test_contents_define_dataset_builder.py (hunk 5 above)
Comment (htu) [reply tests/unit/test_dataset_builders/test_contents_define_dataset_builder.py:367]: Removed the commented variable.

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py (hunk 4 above)
Comment (htu) [reply cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:201]: Removed.

File: cdisc_rules_engine/rules_engine.py (hunk 2 above)
Comment (gerrycampion) [reply cdisc_rules_engine/rules_engine.py:210]: I think this whole `if` branch should be removed. Any necessary logic should be inside the builder instead.
And I don't think we need `define_metadata`

Comment (htu) [reply cdisc_rules_engine/rules_engine.py:210]: Removed the block.

File: cdisc_rules_engine/dataset_builders/base_dataset_builder.py [hunk 6]
```diff
@@ -116,3 +118,14 @@ def get_define_xml_value_level_metadata(self) -> List[dict]:
...
+
+    def get_define_metadata(self):
+        directory_path = get_directory_path(self.dataset_path)
+        define_xml_path: str = os.path.join(directory_path, DEFINE_XML_FILE_NAME)
+        define_xml_contents: bytes = self.data_service.get_define_xml_contents(
+            dataset_name=define_xml_path
+        )
+        define_xml_reader = DefineXMLReaderFactory.from_file_contents(
+            define_xml_contents, cache_service_obj=self.cache
+        )
```
Comment (gerrycampion) [inline cdisc_rules_engine/dataset_builders/base_dataset_builder.py:130]: Can we move this to `get_define_xml_reader` to be shared by all of these methods?

File: cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        define_dataset_label - dataset label from define
+        define_dataset_location - dataset location from define
+        define_dataset_class - dataset class
+        define_dataset_structure - dataset structure
+        define_dataset_is_non_standard - whether a dataset is a standard
+        define_dataset_variables - dataset variable list
+
+        ...,
+        """
+        v_prg = "get_dataset_define_metadata"
```
Comment (gerrycampion) [inline cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py:31]: what does v_prg mean? If you need this for logging, maybe use something like `__class__` or `__class__.__name__` instead

File: cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        # 1. Build define xml dataframe
```
Comment (gerrycampion) [inline cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py:32]: probably good to break these dataframe build steps into separate methods

**Review by gerrycampion — CHANGES_REQUESTED** [review]

File: cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py
[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py:1]: The md was updated.

File: cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py (hunk 8 above)
Comment (htu) [reply cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py:32]: It has been broken up to two separate methods.

File: cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        define_metadata = self.get_define_metadata()
+        define_df = pd.DataFrame(define_metadata)
+
+        if define_df.empty:
+            define_df = pd.DataFrame(columns=define_col_order)
+            logger.info(f"No define_metadata is provided for {__name__}.")
+        return define_df
+
+    def _get_dataset_dataframe(self):
+        v_prg = f"{__name__}::_get_dataset_dataframe"
```
Comment (gerrycampion) [inline cdisc_rules_engine/dataset_builders/contents_define_dataset_builder.py:71]: I still don't know what v_prg means. again, you should use _get_dataset_dataframe.__name__  instead of hardcoding in string. This way, if the method name is ever changed, it will be detected and not get out of sync

File: cdisc_rules_engine/dataset_builders/contents_dataset_builder.py [hunk 10]
```diff
@@ -22,5 +22,5 @@ def get_dataset(self, **kwargs):
             )
         else:
             # single dataset. the most common case
-            dataset: pd.DataFrame = self.build()
+            dataset: pd.DataFrame = self.build(**kwargs)
```
Comment (gerrycampion) [inline cdisc_rules_engine/dataset_builders/contents_dataset_builder.py:25]: unnecessary change

File: cdisc_rules_engine/dataset_builders/base_dataset_builder.py [hunk 11]
```diff
@@ -24,6 +24,7 @@ def __init__(
         dataset_path,
         datasets,
         domain,
+        define_metadata: List[dict] = [],
```
[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/dataset_builders/base_dataset_builder.py:27]: Yes, it is done.

File: cdisc_rules_engine/rules_engine.py [hunk 12]
```diff
@@ -322,7 +329,7 @@ def execute_rule(
         )
         results = []
         run(
-            serialize_rule(rule_copy),  # engine expects a JSON serialized dict
+            serialize_rule(rule),  # engine expects a JSON serialized dict
```
Comment (gerrycampion) [inline cdisc_rules_engine/rules_engine.py:326]: why was this changed?

File: cdisc_rules_engine/rules_engine.py [hunk 13]
```diff
@@ -340,14 +341,20 @@ def get_define_xml_metadata_for_domain(
         """
         Gets Define XML metadata and returns it as dict.
         """
-        directory_path = get_directory_path(dataset_path)
+        if os.path.exists(dataset_path):
```
Comment (gerrycampion) [inline cdisc_rules_engine/rules_engine.py:344]: why was this changed?

File: cdisc_rules_engine/dataset_builders/contents_dataset_builder.py (hunk 10 above)
Comment (htu) [reply cdisc_rules_engine/dataset_builders/contents_dataset_builder.py]: Added it back.

File: cdisc_rules_engine/rules_engine.py (hunk 12 above)
Comment (htu) [reply cdisc_rules_engine/rules_engine.py:326]: Used the copied rule.

File: cdisc_rules_engine/rules_engine.py (hunk 13 above)
Comment (htu) [reply cdisc_rules_engine/rules_engine.py:344]: If the dataset_path is already an exising path, we should be able to use it driectly.

File: cdisc_rules_engine/dataset_builders/base_dataset_builder.py (hunk 6 above)
Comment (htu) [reply cdisc_rules_engine/dataset_builders/base_dataset_builder.py:128]: You mean to move to define_xml_reader_factory.py (DefineXMLReaderFactory)?  I do not see a get_define_xml_reader file. I tried to move it to DefineXMLReaderFactory, it causes 12 tests to be failed. So I am keeping it here.

Comment (gerrycampion) [reply cdisc_rules_engine/dataset_builders/base_dataset_builder.py:128]: No, i mean create a method, `get_define_xml_reader` in this same file that contains this section of the current method. there are several other methods within this same file that can share the same chunk of code.

File: cdisc_rules_engine/rules_engine.py (hunk 13 above)
Comment (gerrycampion) [reply cdisc_rules_engine/rules_engine.py:344]: I don't think that's what we want. get_directory_path gets the directory name. If dataset_path is a file that exists, this added logic will provide the file path instead of directory name.
Ultimately it doesn't make a difference because I think path.join might still work with the file path instead of dir path, but this extra logic appears to be unnecessary and I think it should be removed.

**Review by gerrycampion — CHANGES_REQUESTED** [review]
2 items remaining

File: cdisc_rules_engine/dataset_builders/base_dataset_builder.py (hunk 6 above)
Comment (htu) [reply cdisc_rules_engine/dataset_builders/base_dataset_builder.py:128]: Got it and created it.

File: cdisc_rules_engine/rules_engine.py (hunk 13 above)
Comment (htu) [reply cdisc_rules_engine/rules_engine.py:344]: OK. I removed it.

File: cdisc_rules_engine/dataset_builders/base_dataset_builder.py (hunk 6 above)
Comment (gerrycampion) [reply cdisc_rules_engine/dataset_builders/base_dataset_builder.py]: looks good now, except this new method can also be used by `get_define_xml_item_group_metadata` in addition to the other 2

**Review by gerrycampion — CHANGES_REQUESTED** [review]
one more thing

**Review by gerrycampion — APPROVED** [review]
